utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+	pub fn new(json: JsonValue) -> Option<Self> {
+		if json.as_object().map_or(false, |obj| {
+			obj.contains_key("event") && obj.contains_key("phase") && obj.contains_key("topics")
+		}) {
+			Some(Self(json))
+		} else {
+			None
+		}
+	}
+
+	/// Returns the inner event if it matches the given extrinsic index.
+	pub fn extrinsic_event(&self, extrinsic_index: u64) -> Option<DynamicEvent> {
+		self.0
+			.pointer("/phase/ApplyExtrinsic")
+			.and_then(|index| index.as_u64())
+			.filter(|index| *index == extrinsic_index)
+			.map(|_| self.event())
+	}
+
+	/// Returns the inner event and it's extrinsic index if the event was triggered during the
+	/// ApplyExtrinsic phase.
+	pub fn indexed_extrinsic_event(&self) -> Option<(u64, DynamicEvent)> {
+		self.0
+			.pointer("/phase/ApplyExtrinsic")
+			.and_then(|index| index.as_u64())
+			.map(|index| (index, self.event()))
+	}
+
+	/// Return a clone of the inner event.
+	pub fn event(&self) -> DynamicEvent {
+		DynamicEvent(
+			self.0
+				.pointer("/event")
+				.expect("existence of /event is ensured by constructor")
+				.clone(),
+		)
+	}
+
+	/// Unwrap the inner JSON value.
+	pub fn into_inner(self) -> JsonValue {
+		self.0
+	}
+}
+
+impl DynamicEvent {
+	/// If this event represents the outcome of an extrinsic, return the result.
+	pub fn extrinsic_outcome(&self) -> Option<Result<(), DispatchError>> {
+		if self.0.pointer("/System/ExtrinsicSuccess/dispatch_info").is_some() {
+			Some(Ok(()))
+		} else {
+			self.0
+				.pointer("/System/ExtrinsicFailed/dispatch_error")
+				.map(|dispatch_error_json| {
+					Err((|| {
+						Some(DispatchError::KnownModuleError {
+							pallet: dispatch_error_json["pallet"].as_str()?.to_string(),
+							name: dispatch_error_json["name"].as_str()?.to_string(),
+							error: dispatch_error_json["error"].as_str()?.to_string(),
+						})
+					})()
+					.unwrap_or_else(|| DispatchError::DispatchError(dispatch_error_json.clone())))
+				})
+		}
+	}
+
+	pub fn to_json(self) -> JsonValue {
+		self.0.clone()
+	}
+}
+
+type NamedVariantLookup<T> = BTreeMap<u8, (String, T)>;
+type ErrorInfo = NamedVariantLookup<Vec<String>>;
+
+/// A decoder for substrate events.
+///
+/// Events are decoded dynamically based on metadata provided on construction of this type.
+///
+/// For the System.ExtrinsicFailed event, the dispatch error is resolved to a structured
+/// human-readable error message with the following shape:
+///
+/// ```json
+/// {
+///     "pallet": "PalletName",
+///     "name": "ErrorName",
+///     "error": "Human readable error message"
+/// }
+/// ```
+#[derive(Debug)]
+pub struct EventDecoder {
+	events_type_id: u32,
+	types: PortableRegistry,
+	errors: NamedVariantLookup<ErrorInfo>,
+}
+
+impl EventDecoder {
+	/// Create a new event decoder.
+	///
+	/// The arguments to this constructor can be extracted from substrate metadata.
+	pub fn new(mut types: PortableRegistry, errors_type_id: u32) -> Option<Self> {
```

Comment:
What's the reasoning for requiring this separate type id for the errors? Assuming this is some substrate specific way of doing things? why is that?

Reply:
The errors type id is available from the metadata directly. The metadata contains the type ids for `RuntimeError`, `RuntimeEvent`, `RuntimeCall` but sadly not the `EventRecord<RuntimeEvent>`, which is why the latter needs to be searched for.